callback_test.go

```diff
@@ -0,0 +1,83 @@
+//go:build darwin
+
+package purego_test
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"testing"
+	"unsafe"
+
+	"github.com/ebitengine/purego"
+)
+
+// TestCallGoFromSharedLib is a test that checks for stack corruption on darwin/arm64
+// when C calls Go code from a non-Go thread in a dynamically loaded share library.
+func TestCallGoFromSharedLib(t *testing.T) {
+	libFileName := filepath.Join(t.TempDir(), "libcbtest.so")
+	t.Logf("Build %v", libFileName)
+
+	if err := buildSharedLib(libFileName, "libcbtest/callback.c"); err != nil {
+		t.Fatal(err)
+	}
+	defer os.Remove(libFileName)
+
+	lib, err := purego.Dlopen(libFileName, purego.RTLD_NOW|purego.RTLD_GLOBAL)
+	if err != nil {
+		t.Fatalf("Dlopen(%q) faled: %v", libFileName, err)
+	}
+
+	var initLib func()
+	purego.RegisterLibFunc(&initLib, lib, "initLib")
+	initLib()
+
+	var printString func(p uintptr, s string)
+	purego.RegisterLibFunc(&printString, lib, "printString")
+
+	var wg sync.WaitGroup
+
+	goFunc := func(cstr *byte, n int) int {
+		s := string(unsafe.Slice(cstr, n))
+		t.Logf("FROM Go: %s\n", s)
+		wg.Done()
+		return 1
+	}
+
+	cb := purego.NewCallback(goFunc)
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		printString(cb, "a test string")
+	}
+
+	wg.Wait()
+}
+
+func buildSharedLib(libFile string, sources ...string) error {
+	out, err := exec.Command("go", "env", "-json").Output()
```

Comment:
running `go env CC` and trimming the result by `strings.TrimSpace` should be fine. We don't need parsing JSON.

Reply:
Done